src/window/windows/castlewindow_winapi.inc

```diff
@@ -367,9 +368,31 @@ function TCastleWindowCustom.WindowProc(uMsg: UINT; wParm: WPARAM; lParm: LPARAM
          WM_KEYDOWN message is not translatable as WM_CHAR) OR
       2. we have exactly one WM_CHAR message in our queue, and this is WM_CHAR
          message corresponding to current WM_KEYDOWN message. }
+
     if PeekMessage(@CharMsg, h_Wnd, WM_CHAR, WM_CHAR, PM_REMOVE) then
-      CharKey := Chr(CharMsg.wParam) else
-      CharKey := #0;
+    begin
+      { Now we do some unsafe magic.
+        MultiByteToWideChar converts wParam:longint to a KeyChar:WideChar
+        which should be UTF-16. "Should" doesn't mean it is.
+        First problem here is that not all characters can fit into UTF-16.
+        Second, we suggest that system-wide default encoding CP_ACP is the default
+        for current thread, however CP_THREAD_ACP should be used better,
+        thou it is unavailable in defines.inc }
+      MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, @CharMsg.wParam, SizeOf(CharMsg.wParam), @KeyChar, SizeOf(KeyChar));
+      { Now, hopefully convert WideChar to UTF8Char }
+      KeyString := UTF8Encode(KeyChar);
+      { This is a bugfix with Ukrainian "i" letter which always messes up encodings :)
+        I just replace it by English "i" here,
+        however, I'm really if this is a good idea,
+        as this will make the words spell wrong (English letter instead of Ukrainian)
+        and will mess up with automatic word processing, e.g. spellchecking.
+        It seems that the character actually gets converted correctly, but
+        Windows functions (e.g. output to console window) don't understand it }
+      {if CharMsg.wParam = 179 then
+        KeyString := 'i';}
+    end
```

Comment:
This comment (from _"{ This is a bugfix with Ukrainian "i"..."_) can be removed :) As far as I understand, it works correctly in Castle Game Engine?

(And indeed, the Windows console is another beast. It does not expect strings in UTF-8 encoding. This is unrelated to CGE, you simply cannot directly output UTF-8 string on a Windows console. I saw a thread on Lazarus mailing list about it, but I did not follow... in any case, this is outside of CGE. We want strings in UTF-8, just like Lazarus. And our log wants to output UTF-8, because it may go to file, and writing a text file with UTF-8 encoding is fine and desired.)

Reply:
Yes, I think we can remove this comment, as Ukrainian "i" gets converted to UTF-8 correctly. It's Windows console problem :) Done.